### PR TITLE
Tpu trainer

### PR DESCRIPTION
**Pull request by LysandreJik**
This PR aims to bring TPU support to the trainer. It runs on GLUE/MRPC but is yet untested on others.

Left to do:

- [ ] Saving and reloading mid-training
- [ ] Check it runs on a few examples (Language modeling, NER, other GLUE tasks)
- [ ] Write down training speed-ups
- [ ] Write down evaluation speed-ups

The API is not final.

### CODE REVIEW

File: examples/run_glue.py [hunk 1]
```diff
@@ -202,5 +190,31 @@ def compute_metrics(p: EvalPrediction) -> Dict:
...
+        model_args, data_args, training_args = parser.parse_args_into_dataclasses()
+
+    if training_args.use_tpu:
+        import torch_xla.distributed.xla_multiprocessing as xmp
+
+        return xmp.spawn(_mp_fn, args=(model_args, data_args, training_args), nprocs=training_args.num_cores)
+    else:
+        return run_glue(model_args, data_args, training_args)
+
+
```
Comment (LysandreJik) [inline examples/run_glue.py:218]: This seems like something we would need to put in every script, so there's probably a way to cleanly factor this in the trainer.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -262,16 +304,21 @@ def train(self, model_path: Optional[str] = None):
...
+            if is_tpu_available
+            else (
+                self.args.train_batch_size
+                * self.args.gradient_accumulation_steps
+                * (torch.distributed.get_world_size() if self.args.local_rank != -1 else 1),
+            )
+        )
         logger.info("***** Running training *****")
-        logger.info("  Num examples = %d", len(train_dataloader.dataset))
+        logger.info("  Num examples = %d", len(train_dataloader) * self.args.train_batch_size)
```
Comment (LysandreJik) [inline src/transformers/trainer.py:318]: This was crashing

File: examples/run_glue.py [hunk 3]
```diff
@@ -61,20 +61,7 @@ class ModelArguments:
...
-    parser = HfArgumentParser((ModelArguments, DataTrainingArguments, TrainingArguments))
-
-    if len(sys.argv) == 2 and sys.argv[1].endswith(".json"):
-        # If we pass only one argument to the script and it's the path to a json file,
-        # let's parse it to get our arguments.
-        model_args, data_args, training_args = parser.parse_json_file(json_file=os.path.abspath(sys.argv[1]))
-    else:
-        model_args, data_args, training_args = parser.parse_args_into_dataclasses()
-
+def run_glue(model_args, data_args, training_args):
```
Comment (LysandreJik) [inline examples/run_glue.py:64]: I put `run_glue` here as a placeholder, but could be replaced by something better like `run()` or something else.

I chose to keep `main()` as the method that can launch the full training, as it is what we call from tests.

File: src/transformers/training_args.py [hunk 4]
```diff
@@ -94,6 +106,9 @@ class TrainingArguments:
         },
     )
     local_rank: int = field(default=-1, metadata={"help": "For distributed training: local_rank"})
+    use_tpu: bool = field(default=False, metadata={"help": "Whether to use a TPU device."})
+    num_cores: int = field(default=8, metadata={"help": "Number of TPU cores to use (1 or 8)."})
```
Comment (LysandreJik) [inline src/transformers/training_args.py:110]:
```suggestion
    num_tpu_cores: int = field(default=8, metadata={"help": "Number of TPU cores to use (1 or 8)."})
```

**Comment by thomwolf** [comment]
Awesome!

**Pull request merged**